### DISCUSSION TITLE
Help in making sure I am using this crate right in my project

### DISCUSSION THREAD

**Discussion by conner-replogle** [rpgp/rpgp#423] [Q&A]
[Github Repo](https://github.com/conner-replogle/blackbox)

I did a 24 hour run to make this application for practice to fix a issue I had with pgp apps on mac os. While it works for decrypting and encrypting in my use cases, I think the project turned out nice enough to try and make it available to all.  However I have no Idea if I did anything right and I mostly just looked through the source and try to figure out how to do stuff. (I did not find the usage examples in docs till I got to the key gen lol). So if anyone wants to look through the decrypt and encrypt for me under the src-tauri  folder that would be very helpful. Mostly I have issues with the parameters to some functions for encrypting as I picked the most recognizable options for me I don't know the standard. Also if anyone has recommendations for what I should add also let me know.

**Comment by hko-s** [comment]
Hey Conner, thanks for getting in touch and sharing your project!
I just had a quick peek, but won't be able to look at everything in detail, right now.

Still, I wanted to share two pointers:

1. Key generation

When generating new OpenPGP keys, it's currently most common to use ECC algorithms. Additionally, it's common in OpenPGP to have "subkeys" that get used for specific purposes. This adds complexity, but is a best practice.

Roughly like this (some details are not fleshed out here, like the preferred algorithms, which can be left unset, but are nice to set):

```rust
        let subkeys = vec![SubkeyParamsBuilder::default()
                .key_type(pgp::KeyType::ECDH(ECCCurve::Curve25519))
                .can_encrypt(true)
                .build()?];

        let mut key_params = SecretKeyParamsBuilder::default();
        key_params
            .key_type(pgp::KeyType::EdDSALegacy)
            .can_certify(true)
            .can_sign(true)
            .primary_user_id(primary_user_id)
            .preferred_symmetric_algorithms(PREFERRED_SYMMETRIC_KEY_ALGORITHMS.into())
            .preferred_hash_algorithms(PREFERRED_HASH_ALGORITHMS.into())
            .preferred_compression_algorithms(PREFERRED_COMPRESSION_ALGORITHMS.into())
            .subkeys(subkeys);

        let secret_key_params = key_params.build()?;
        let secret_key = secret_key_params.generate(&mut rng)?;

        let signed_secret_key = secret_key.sign(&mut thread_rng(), String::new)?;
```

There's a lot more to think about, such as: the private key material can be protected with a password. Depending on the use case, this may be beneficial to use.

Also, when decrypting, you need to take care to use the encryption subkey.

2. SEIPDv1 vs. SEIPDv2

I noticed you were unsure about these two encryption mechanisms. And that's more than fair. SEIPDv2 is a new mechanism that was introduced in the very new RFC 9580. If you're aiming for broad compatibility, it's way too early to use SEIPDv2, so I'd say you made the right call in using SEIPDv1 encryption.


**Some more general remarks**

OpenPGP is a pretty deep can of worms, and rPGP is a very low level abstraction. Making your application truly "watertight" would take quite a bit of thought and work. For example, subkeys can be revoked, and a cautious sender will ideally check if their recipient has marked their encryption subkey as "revoked", or if the subkey has expired because it is too old.
Implementing all of this on the basis of rPGP is a lot of work, and - I think - puts too much of an onus on application developers like you.

I hope that we will be able to offer an API that exposes abstractions that are more easy to use, in applications like yours.
If you're interested in testing such an API, I'd ping you in a while, if there is an update on that front :)

Either way, take care, and feel free to send more questions, if you continue to delve into OpenPGP!

If you do want to dig deeper, I'm happy to offer more pointers, and spend some more time reviewing your approach :)

**OpenPGP book**

Finally, if you'd like to read up on OpenPGP, you might enjoy https://openpgp.dev/
